packages/datadog-plugin-mongodb-core/src/index.js

```diff
@@ -1,6 +1,121 @@
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+class MongodbCorePlugin extends Plugin {
+  static get name () {
+    return 'mongodb-core'
+  }
 
-module.exports = [].concat(unified, legacy)
+  addSubs (func, start, asyncEnd = defaultAsyncEnd) {
+    this.addSub(`apm:mongodb:${func}:start`, start)
+    this.addSub(`apm:mongodb:${func}:end`, this.exit.bind(this))
+    this.addSub(`apm:mongodb:${func}:error`, errorHandler)
+    this.addSub(`apm:mongodb:${func}:async-end`, asyncEnd)
+  }
+
+  startSpan (customTags, store, ns, cmd, topology, operationName) {
+    const query = getQuery(cmd)
+    const resource = getResource(ns, query, operationName)
+    const tags = {
+      'service.name': this.config.service || `${this.tracer._service}-mongodb`,
+      'resource.name': resource,
+      'span.type': 'mongodb',
+      'span.kind': 'client',
+      'db.name': ns
+    }
+
+    for (const tag in customTags) {
+      tags[tag] = customTags[tag]
+    }
+    const span = this.tracer.startSpan('mongodb.query', {
+      childOf: store ? store.span : null,
+      tags
+    })
+
+    if (query) {
+      span.setTag('mongodb.query', query)
+    }
+
+    const options = topology && topology.s && topology.s.options
+
+    if (options && options.host && options.port) {
+      span.addTags({
+        'out.host': topology.s.options.host,
+        'out.port': topology.s.options.port
+      })
+    }
+
+    analyticsSampler.sample(span, this.config.measured)
+    return span
+  }
+
+  constructor (...args) {
+    super(...args)
+
+    this._addCommandSubs('command')
+    this._addCommandSubs('insert')
+    this._addCommandSubs('update')
+    this._addCommandSubs('remove')
+    this._addCommandSubs('getMore')
+    this._addCommandSubs('query')
+    this._addCommandSubs('_find')
+    this._addCommandSubs('killCursors')
+  }
+  _addCommandSubs (name) {
+    this.addSubs(name, ({ ns, ops, server, name }) => {
+      const store = storage.getStore()
+      const span = this.startSpan({}, store, ns, ops, server, name)
+      this.enter(span, store)
+    })
+  }
+}
+
+function defaultAsyncEnd () {
+  storage.getStore().span.finish()
+}
+
+function errorHandler (err) {
+  storage.getStore().span.setTag('error', err)
+}
+
+function getQuery (cmd) {
+  if (!cmd || typeof cmd !== 'object' || Array.isArray(cmd)) return
+  if (cmd.query) return JSON.stringify(sanitize(cmd.query))
+  if (cmd.filter) return JSON.stringify(sanitize(cmd.filter))
+}
+
+function getResource (ns, query, operationName) {
+  const parts = [operationName, ns]
+
+  if (query) {
+    parts.push(query)
+  }
+
+  return parts.join(' ')
+}
+
+function sanitize (input) {
+  const output = {}
+
+  if (!isObject(input) || Buffer.isBuffer(input) || isBSON(input)) return '?'
+
+  for (const key in input) {
+    if (typeof input[key] === 'function') continue
+
+    output[key] = sanitize(input[key])
+  }
+
+  return output
+}
+
+function isObject (val) {
```

Comment:
Do we not have some global helper for this?

Reply:
Not currently, should I create one?

Reply:
This is somewhat specific, because it checks that it's an object _and_ not an array.